### PR TITLE
NXP-30814: handle nested compound documents

### PR DESCRIPTION
**Pull request by NourNuxeo**

### CODE REVIEW

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 1]
```diff
@@ -106,7 +109,19 @@ protected void createEntry(DocumentModel compoundDoc, ZipFile zip, ZipEntry entr
             var doc = session.createDocumentModel(parentDocPath, entryDocName, compoundFolderType);
             session.createDocument(doc);
         } else {
-            var blob = new ZipEntryBlob(zip, entry);
+            Blob blob;
+            if (entry.getName().endsWith(".zip")) {
```
Comment (NourNuxeo) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: I think we can make something more resilient for type checking, like a regex so the case is not an issue.
Maybe if/when the time to accept multiple extensions comes?

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 2]
```diff
@@ -69,12 +71,24 @@ public DocumentModel createCompoundDocument(CoreSession session, String parent,
...
             return finalDoc;
         } catch (IOException e) {
-            var message = String.format("Failed to create CompoundDocument for archive: %s in parent: %s",
+            var message = String.format("Failed to create compound document for archive: %s in parent: %s",
                     archive.getFilename(), parent);
             throw new NuxeoException(message, e);
         }
     }
 
+    protected File getZipFile(Blob archive) {
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: Could you rename the method to `getFile` because the method returns a `File`, and `new ZipFile(getZipFile)` could be strange.

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 3]
```diff
@@ -108,19 +122,25 @@ protected void createEntry(DocumentModel compoundDoc, ZipFile zip, ZipEntry entr
...
+            Framework.getService(FileManager.class).createOrUpdateDocument(ctx);
+        } catch (NuxeoException | IOException e) {
+            String message = String.format("Failed to create document for entry: %s in: %s for archive: %s", entry,
+                    compoundDoc, zip);
+            if (e instanceof NuxeoException) {
+                NuxeoException ne = (NuxeoException) e;
+                ne.addInfo(message);
+                throw ne;
             }
+            throw new NuxeoException(message, e);
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: Is this change needed?

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/CompoundDocumentUtils.java [hunk 4]
```diff
@@ -44,12 +44,21 @@ private CompoundDocumentUtils() {
...
                 CompoundDocumentUtils.class.getResource("/files/defaultCompound").getPath()).listFiles();
+        return getArchive("test.zip", sources);
+    }
+
+    protected static Blob getNestedTestArchives() throws IOException {
+        File source = getTestArchive().getFile();
+        return getArchive("nest.zip", source);
+    }
+
+    protected static Blob getArchive(String filename, File... source) throws IOException {
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/CompoundDocumentUtils.java]:
```suggestion
    protected static Blob getArchive(String filename, File... sources) throws IOException {
```

File: nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java [hunk 5]
```diff
@@ -123,4 +124,18 @@ public void testBadCustomCompoundDocumentFolderType() throws IOException {
         assertEquals("Failed to invoke operation " + COMPOUND_FOLDER_DOCTYPE_DETECTION_OPERATION, cause.getMessage());
     }
 
+    @Test
+    public void testCreateNestedCompoundDocuments() throws IOException {
+        Blob blob = CompoundDocumentUtils.getNestedTestArchives();
```
Comment (kevinleturc) [inline nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentImporter.java]: Should `getNestedTestArchives` be imported statically to follow the other methods?

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 6]
```diff
@@ -108,12 +122,13 @@ protected void createEntry(DocumentModel compoundDoc, ZipFile zip, ZipEntry entr
         } else {
             var blob = new ZipEntryBlob(zip, entry);
             blob.setFilename(entryDocName);
-            FileImporterContext ctx = FileImporterContext.builder(session, blob, parentDocPath).build();
+            FileImporterContext ctx = FileImporterContext.builder(compoundDoc.getCoreSession(), blob, parentDocPath)
+                                                         .build();
```
Comment (troger) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: why the change here?

**Review by kevinleturc — DISMISSED** [review]

File: nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java [hunk 7]
```diff
@@ -108,12 +122,13 @@ protected void createEntry(DocumentModel compoundDoc, ZipFile zip, ZipEntry entr
         } else {
             var blob = new ZipEntryBlob(zip, entry);
             blob.setFilename(entryDocName);
-            FileImporterContext ctx = FileImporterContext.builder(session, blob, parentDocPath).build();
+            FileImporterContext ctx = FileImporterContext.builder(compoundDoc.getCoreSession(), blob, parentDocPath)
```
Comment (guirenard) [inline nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java]: I would have passed as a param of the `createEntry` method from the caller `createCompoundDocument` and reused it here. It is better to make sure we always use the same session object (in case of unrestricted runner, detached docs, etc.)

**Review by kevinleturc — APPROVED** [review]

**Review by troger — APPROVED** [review]
